### PR TITLE
Improve utility calculation

### PR DESCRIPTION
**Pull request by mdubied**
# Summary
Implement an improved version of utility (resolves #270)

## Additional Information
@JaafarRammal the global calculations that should be done on frontend are:
1) utility by agent team over time : sum over agents of one team (utility) / # team agents
2) global utility over time: sum over all agent teams (utility) / # agent teams

### CODE REVIEW

File: pkg/infra/baseAgent.go [hunk 1]
```diff
@@ -262,75 +265,42 @@ func (a *Base) UpdateHPChange(change int) {
...
+	beta := 0.1
+	gamma := 0.18
+
+	g := float64(a.CurrPlatFood()) / float64(a.tower.maxPlatFood)
+	var q float64
+	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
+		q = 1.0
+	} else {
+		q = 0.0
 	}
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go]:
```suggestion
	resNeed := 0.0
	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
		resNeed = 1.0
	}
```

File: pkg/infra/baseAgent.go [hunk 2]
```diff
@@ -262,75 +265,42 @@ func (a *Base) UpdateHPChange(change int) {
...
-	return float64(intake) / float64(seen)
-}
+	p := 0.0
+	r_prime := float64(foodTaken) / float64(a.tower.healthInfo.MaxFoodIntake)
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:305]:
```suggestion
	resApprop := float64(foodTaken) / float64(a.tower.healthInfo.MaxFoodIntake)
```

File: pkg/infra/baseAgent.go [hunk 3]
```diff
@@ -262,75 +265,42 @@ func (a *Base) UpdateHPChange(change int) {
...
+// Which provides the utility of an agent, ui, as:
+// ui =
+// αi(qi) + βi(Ri −qi), if Ri ≥qi
+// αi(Ri) −γi(qi −Ri), otherwise
+func (a *Base) utility(foodRequested, foodTaken food.FoodType) {
+	alpha := 0.2
+	beta := 0.1
+	gamma := 0.18
+
+	g := float64(a.CurrPlatFood()) / float64(a.tower.maxPlatFood)
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:291]:
```suggestion
	resAvail := float64(a.CurrPlatFood()) / float64(a.tower.maxPlatFood)
```

File: pkg/infra/baseAgent.go [hunk 4]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
+
+	// If the agent is on the verge of dying, food is needed
+	var q float64
+	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
+		q = 1.0
+		// If an agent is not about to die, it effectively needs 0 food
+	} else {
+		q = 0.0
+	}
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:300]:
```suggestion
	// If an agent is not about to die, it effectively needs 0 food
	resNeed := 0.0
	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
		resNeed = 1.0
	}
```

File: pkg/infra/baseAgent.go [hunk 5]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
+	// No resources are recontributed to the common pool (provision = 0)
+	p := 0.0
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:302]:
```suggestion
	provision := 0.0
```

File: pkg/infra/baseAgent.go [hunk 6]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
-	return simpleQuadraticUtility(hpGainFactor * foodIntakeFactor)
-}
+// Updates the utility of a given agent:
+// We take N agents i ∈{1,...,N}that perform the following actions in each iterated round t ∈{1,...,∞}:
+// 1.) Determines the resources it has available, gi ∈[0,1]
+// 2.) Determines its need for resources, qi ∈[0,1]
+// 3.) Makes a provision of resources, pi ∈[0,1]
+// 4.) Makes a demand for resources, di ∈[0,1]
+// 5.) Receives an allocation of resources, ri ∈[0,1]
+// 6.0 Makes an appropriation of resources, r′i ∈[0,1]
```
Comment (MattScottEEE) [inline pkg/infra/baseAgent.go:275]:
```suggestion
// 6.) Makes an appropriation of resources, r′i ∈[0,1]
```
Typo

File: pkg/infra/baseAgent.go [hunk 7]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
-}
+	// Note that r isn't included. This is because it is just a metric for assessing Ostrom Institution Theory
+	// r >= r_prime (for conventional institution)
 
-// Quadratic function standard form: a(x-h)^2+k
-func quadratic(x, a, h, k float64) float64 {
-	return a*math.Pow((x-h), 2) + k
-}
+	// Resource generation by agent per day is what's available, minus what they give back, plus what they actually take
+	R := r_prime + (g - p)
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:311]:
```suggestion
	resAccrued := resApprop + (resAvail - prov)
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: pkg/infra/baseAgent.go [hunk 8]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
-*/
-func foodIntakeFactor(intake, seen food.FoodType) float64 {
-	// If an agent hasn't seen any food, has 0 utility
-	if seen == 0 {
-		return 0
-	}
-	return float64(intake) / float64(seen)
-}
+	// Note that r isn't included. This is because it is just a metric for assessing Ostrom Institution Theory
+	// r >= r_prime (for conventional institution)
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:308]: Irrelevant here. Put it in the report if you want.

File: pkg/infra/baseAgent.go [hunk 9]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
 
-// Quadratic function standard form: a(x-h)^2+k
-func quadratic(x, a, h, k float64) float64 {
-	return a*math.Pow((x-h), 2) + k
-}
+	// Resource generation by agent per day is what's available, minus what they give back, plus what they actually take
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:310]: Would order this so it is in the same order as the formula.

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: pkg/infra/baseAgent.go (hunk 8 above)
Comment (MattScottEEE) [reply pkg/infra/baseAgent.go:308]: I think it's important. It's a variable that's included inside of the function description that magically doesn't appear. Having this comment explicitly shows that it's a utility calculation, not a validity calculation

File: pkg/infra/baseAgent.go [hunk 10]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
+// 6.) Makes an appropriation of resources, r′i ∈[0,1]
+
+// The total resources accrued at the end of a round is hence given by:
+// Ri = r′i + (gi −pi)
+// Which provides the utility of an agent, ui, as:
+// ui = αi(qi) + βi(Ri −qi), if Ri ≥qi
+// 		αi(Ri) −γi(qi −Ri), otherwise
+func (a *Base) utility(foodRequested, foodTaken food.FoodType) {
+	// Hyperparameters that are chosen to make utility commensurate for all agents
+	// These initialisations are chosen by Pitt, but may require tuning
```
[the remark this answers is not in the document]
Comment (MattScottEEE) [reply pkg/infra/baseAgent.go:284]: Fine to delete, but should be made known to the cohort

File: pkg/infra/baseAgent.go (hunk 1 above)
Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: Fair.

Since this formula comes direct from literature (and is explicitly defined above), I would rather keep the variable names as letters

File: pkg/infra/baseAgent.go (hunk 8 above)
Comment (jzzheng22) [reply pkg/infra/baseAgent.go:308]: I wouldn't put it in the function description then.

File: pkg/infra/baseAgent.go [hunk 11]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
+// 4.) Makes a demand for resources, di ∈[0,1]
+// 5.) Receives an allocation of resources, ri ∈[0,1]
+// 6.) Makes an appropriation of resources, r′i ∈[0,1]
+
+// The total resources accrued at the end of a round is hence given by:
+// Ri = r′i + (gi −pi)
+// Which provides the utility of an agent, ui, as:
+// ui = αi(qi) + βi(Ri −qi), if Ri ≥qi
+// 		αi(Ri) −γi(qi −Ri), otherwise
+func (a *Base) utility(foodRequested, foodTaken food.FoodType) {
```
Comment (JaafarRammal) [inline pkg/infra/baseAgent.go:282]:
```suggestion
func (a *Base) updateUtility(foodRequested, foodTaken food.FoodType) {
```

File: pkg/infra/baseAgent.go (hunk 1 above)
Comment (jzzheng22) [reply pkg/infra/baseAgent.go]: It's more readable this way - report can use the letters, but here it's quite obscure.

File: pkg/infra/baseAgent.go [hunk 12]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
-}
+	// Resource generation by agent per day is what's available, minus what they give back, plus what they actually take
+	R := r_prime + (g - p)
 
-// Utility function is a strictly increasing function that starts at 0,0 ends at 1,1
-func simpleQuadraticUtility(x float64) float64 {
-	// -(x-1)^2 + 1
-	return utilFunctions.RestrictToRange(0, 1, quadratic(x, -1, 1, 1))
+	if R >= q {
+		a.currentUtility = alpha*q + beta*(R-q)
```
Comment (JaafarRammal) [inline pkg/infra/baseAgent.go:314]:
```suggestion
		a.utility = alpha*q + beta*(R-q)
```

File: pkg/infra/baseAgent.go [hunk 13]
```diff
@@ -262,75 +265,56 @@ func (a *Base) UpdateHPChange(change int) {
...
+	} else {
+		a.currentUtility = alpha*R - gamma*(q-R)
```
Comment (JaafarRammal) [inline pkg/infra/baseAgent.go:316]:
```suggestion
		a.utility = alpha*R - gamma*(q-R)
```

File: pkg/infra/baseAgent.go [hunk 14]
```diff
@@ -344,7 +328,7 @@ func (a *Base) storyState() logging.AgentState {
 		Floor:     a.floor,
 		Age:       a.age,
 		Custom:    "",
-		Utility:   a.utility(),
+		Utility:   a.currentUtility,
```
Comment (JaafarRammal) [inline pkg/infra/baseAgent.go:331]:
```suggestion
		Utility:   a.utility,
```

File: pkg/infra/tower.go [hunk 15]
```diff
@@ -97,14 +97,15 @@ func (t *Tower) Reshuffle() {
...
 
 func (t *Tower) endOfDay() {
-	t.dayInfo.CurrDay += 1
 	for _, agent := range t.Agents {
 		agent := agent.BaseAgent()
 		agent.hpDecay(t.healthInfo)
 		agent.increaseAge()
 		agent.updateTreaties()
-		t.stateLog.LogUtility(t.dayInfo, agent.agentType, agent.utility(), agent.IsAlive())
+		t.stateLog.LogUtility(t.dayInfo, agent.agentType, agent.currentUtility, agent.IsAlive())
```
Comment (JaafarRammal) [inline pkg/infra/tower.go:105]:
```suggestion
		t.stateLog.LogUtility(t.dayInfo, agent.agentType, agent.utility, agent.IsAlive())
```

File: pkg/infra/tower.go [hunk 16]
```diff
@@ -97,14 +97,15 @@ func (t *Tower) Reshuffle() {
...
+		agent.currentUtility = 0
```
Comment (JaafarRammal) [inline pkg/infra/tower.go:106]:
```suggestion
		agent.utility = 0
```
Can have an `EndOfDay` function in agents and we move all of these state updates to it, although can be a different PR

File: pkg/infra/baseAgent.go (hunk 1 above)
Comment (mdubied) [reply pkg/infra/baseAgent.go]: If we use specific variable names in the report, we should probably use the same ones in the code if possible (especially if the formulas are taken from literature). I would rather keep it as it is

Comment (JaafarRammal) [reply pkg/infra/baseAgent.go]: I agree with Jason. Formulas define the concept, but it's better to set readable variable names in the code so the logic can be easily evaluated. E.g: if `q` represents allocation in the formula, it is easier to use a variable name like `foodAlloc` rather than `q`, so when someone is reviewing the code, they can quickly go "aha,`foodAlloc` is being computed correctly" or "hmm I think `foodAlloc` can be obtained in a better way"

File: pkg/infra/tower.go (hunk 16 above)
Comment (MattScottEEE) [reply pkg/infra/tower.go]: Would be nice with a separate PR if ok

File: pkg/infra/baseAgent.go [hunk 17]
```diff
@@ -262,75 +265,58 @@ func (a *Base) UpdateHPChange(change int) {
 	}
 }
 
-func (a *Base) utility() float64 {
-	// TODO: Improve utility calculation
-	// Currently just use ratio between food taken and food seen
-	// Somehow also factor
-	// Ideas: piecewise/bounded function with log, concave quadratic
+// Updates the utility of a given agent:
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:268]:
```suggestion
```

File: pkg/infra/baseAgent.go [hunk 18]
```diff
@@ -262,75 +265,58 @@ func (a *Base) UpdateHPChange(change int) {
...
+func (a *Base) updateUtility(foodRequested, foodTaken food.FoodType) {
+	// We take N agents i ∈{1,...,N}that perform the following actions in each iterated round t ∈{1,...,∞}:
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:270]:
```suggestion
```

File: pkg/infra/baseAgent.go [hunk 19]
```diff
@@ -262,75 +265,58 @@ func (a *Base) UpdateHPChange(change int) {
...
+	// 1.) Determines the resources it has available, gi ∈[0,1]
+	// 2.) Determines its need for resources, qi ∈[0,1]
+	// 3.) Makes a provision of resources, pi ∈[0,1]
+	// 4.) Makes a demand for resources, di ∈[0,1]
+	// 5.) Receives an allocation of resources, ri ∈[0,1]
+	// 6.) Makes an appropriation of resources, r′i ∈[0,1]
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:276]: I mean put each line above the stage where it is relevant, so move the comment 1.) to above the line where `resAvail` is declared, and so on.

Don't need comment 5.

File: pkg/infra/baseAgent.go [hunk 20]
```diff
@@ -262,75 +265,58 @@ func (a *Base) UpdateHPChange(change int) {
...
+	beta := 0.1
+	// The weighting assigned to not getting the resources that are needed
+	gamma := 0.18
+
+	// Available resources normalised by total possible available resources
+	g := float64(a.CurrPlatFood()) / float64(a.tower.maxPlatFood)
+
+	// If an agent is not about to die, it effectively needs 0 food
+	q := 0.0
+	// If the agent is on the verge of dying, food is needed
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:299]:
```suggestion
	// If an agent is not about to die, it effectively needs 0 food
	// If the agent is on the verge of dying, food is needed
	q := 0.0
```

File: pkg/infra/baseAgent.go [hunk 21]
```diff
@@ -262,75 +265,58 @@ func (a *Base) UpdateHPChange(change int) {
...
+	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
+		q = 1.0
+	}
+	// No resources are recontributed to the common pool (provision = 0)
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:303]:
```suggestion
	}
	
	// No resources are recontributed to the common pool (provision = 0)
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: pkg/infra/baseAgent.go (hunk 1 above)
Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: We can do it this way :)

I'll push with better variables

File: pkg/infra/baseAgent.go [hunk 22]
```diff
@@ -82,6 +83,7 @@ func NewBaseAgent(world world.World, agentType agent.AgentType, agentHP int, age
 		totalFoodSeen:  0,
 		totalHPGained:  0,
 		totalHPLost:    0,
+		utility:        0, //discuss initialization
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:86]:
```suggestion
		utility:        0,
```

File: pkg/infra/baseAgent.go [hunk 23]
```diff
@@ -262,75 +265,57 @@ func (a *Base) UpdateHPChange(change int) {
...
+	
+	// No resources are recontributed to the common pool (provision = 0)
+	foodProvision := 0.0
 
-	hpGainFactor := netHPGainFactor(a.totalHPGained, a.totalHPLost, a.tower.healthInfo.MaxHP)
-	foodIntakeFactor := foodIntakeFactor(a.totalFoodTaken, a.totalFoodSeen)
-	// this means the multiplcation of these values can be larger than 1
-	return simpleQuadraticUtility(hpGainFactor * foodIntakeFactor)
-}
+	// Appropriation is the amount of food an agent actually takes, normalised by the max possible
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:305]:
```suggestion
	// Amount of food taken, normalised by the max possible intake
```

File: pkg/infra/baseAgent.go (hunk 20 above)
Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: This refers to the if case, hence q = 1

File: pkg/infra/baseAgent.go [hunk 24]
```diff
@@ -262,75 +265,57 @@ func (a *Base) UpdateHPChange(change int) {
...
-Net HP Gain Factor.
-This one uses net hp gain as opposed to hp gain/loss ratio.
-*/
-func netHPGainFactor(gain, loss, maxhp int) float64 {
-	netHPGain := gain - loss
-	multiplier := 1 + float64(netHPGain)/float64(maxhp) // reward positive net hp gain, punish otherwise
-	return utilFunctions.RestrictToRange(0, math.Inf(1), multiplier)
-}
+	// Note that r isn't included. This is because it is just a metric for assessing Ostrom Institution Theory
+	// r >= r_prime (for conventional institution)
```
Comment (jzzheng22) [inline pkg/infra/baseAgent.go:309]:
```suggestion
```

**Review by jzzheng22 — DISMISSED** [review]

File: pkg/infra/baseAgent.go [hunk 25]
```diff
@@ -262,75 +265,57 @@ func (a *Base) UpdateHPChange(change int) {
...
+	// Available resources normalised by total possible available resources
+	foodAvailable := float64(a.CurrPlatFood()) / float64(a.tower.maxPlatFood)
+
+	// If an agent is not about to die, it effectively needs 0 food
+	foodToSurvive := 0.0
+	// If the agent is on the verge of dying, food is needed
+	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
+		foodToSurvive = 1.0
+	}
+
```
Comment (woonmoon) [inline pkg/infra/baseAgent.go:301]: Is this strictly the case? Surely when it's just critical and not just on its dying day it can count as needing food?

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: But you can still survive. The only time you literally need food is when you're critical and about to die

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: Unless I'm mistaken - can you be critical and die, even if it's not the dying day?

Comment (mdubied) [reply pkg/infra/baseAgent.go]: I would say it is only "needed" the very last day. This is also the way to match the specs, i.e., having agents oscillating between critical state and WeakLevel while spending more time in the critical state.

As a comment, note also that 1.0 is not really the food to survive, but rather ressourceNeeded/maxRessourceNeeded, so that we have q_i in [0,1]. The actual amount of food to survive is actually 2 with the current health parameters

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: Hm, you might be right on this one actually - any food here would be good utility

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: @mdubied I'm happy to go with Gowoon on this one?

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: > I would say it is only "needed" the very last day. This is also the way to match the specs, i.e., having agents oscillating between critical state and WeakLevel while spending more time in the critical state.
> 
> The variable name foodToSurvive is actually not perfect (but ok I think)
> 
> 1.0 is not really the food to survive, but rather ressourceNeeded/maxRessourceNeeded, so that we have q_i in [0,1]. The actual amount of food to survive is actually 2 with the current health parameters

My only issue with this (that makes me want to go with Gowoon) is that taking 2 food while in critical would otherwise be seen as wasted utility. I think this is probably unfair, even though they can technically survive a bit longer. Agents shouldn't be penalised for returning to weak when in critical

Comment (mdubied) [reply pkg/infra/baseAgent.go]: @MattScottEEE you can not die if it's not "the dying day" (what a sentence hahah)

I am not completely sure about it. I completely agree that the food at critical is good utility. But it we push the reasoning, this same amount of food is also good when we are at WeakLevel. With the same amount of food you can either go from ciritical to weak, or stay at weak if you already at weak. Therefore, I would say that the food is really "needed" only to avoid dying, i.e., when `DaysAtCritical=MaxDayCritical`

Comment (woonmoon) [reply pkg/infra/baseAgent.go]: If you're on day 2 at a critical state, you skip eating because day 3 is the last possible day and the platform gets to you on the next day (day 3) and it's empty (probably not an unusual case) - you're dead, you should have eaten on day 2 and you needed that food when it got to you.
This isn't exact because there are no guarantees but I think the critical cut-off is a good place to put the threshold and a useful distinction against idk `weak` for example.

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: Hm, overall I think that we should be rewarding agents for taking food at the critical stage

How about a compromise:

Taking food when at the very final day rewards a q=1.0
Taking food when just at critical (before final day) rewards 0.5<q<1.0

@woonmoon @mdubied

Comment (MattScottEEE) [reply pkg/infra/baseAgent.go]: > If you're on day 2 at a critical state, you skip eating because day 3 is the last possible day and the platform gets to you on the next day (day 3) and it's empty (probably not an unusual case) - you're dead, you should have eaten on day 2 and you needed that food when it got to you.
> This isn't exact because there are no guarantees but I think the critical cut-off is a good place to put the threshold and a useful distinction against idk `weak` for example.

Whilst you may die from this, that doesn't really matter, as you'd technically be getting more utility for your food if you waited another day. It's just unlucky. At the same time, you definitely should get some utility for even waiting until critical. We could parameterise q as a mapping from days critical to 0->1 (daysCrit / maxDaysCrit?)

Comment (mdubied) [reply pkg/infra/baseAgent.go]: It is really a good question to answer @woonmoon !
The argument you givis not completely satisfying in my opinion. Using the same reasoning, I could say that you need to eat already even earlier, for example when HP=20 to avoid the risky scenario where you get nothing on day 3 of being at critical. But we agree that eating when HP = 20 is not "needed"
 I quite like your compromise @MattScottEEE though

Comment (woonmoon) [reply pkg/infra/baseAgent.go]: @mdubied Yes but isn't this why we categorized HP into "how desperately do you need to eat"? Yes you could be unlucky and the platform could arrive empty at you for 20 days in a row and you're just dead. It's just not good utility for an agent to eat when they're healthy but the same can't be said if you know you're ~3 days out from dying. I agree with @MattScottEEE that we should not be penalizing agents for eating when they're critical.

**Review by JaafarRammal — APPROVED** [review]
LGTM and tested it
FSA, utility log size increases linearly with number of agents and number of days (each agent gets its utility logged once per day)

**Review by JaafarRammal — APPROVED** [review]

**Review by daryllimyt — APPROVED** [review]
LGTM Thanks for this

**Review by woonmoon — APPROVED** [review]
Thanks for making the change guys.
